client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -259,7 +259,7 @@ public SendShuffleDataResult sendShuffleData(String appId, List<ShuffleBlockInfo
     }
 
     // maintain the count of blocks that have been sent to the server
-    Map<Long, AtomicInteger> blockIdsTracker = Maps.newConcurrentMap();
+    Map<Long, AtomicInteger> blockIdsTracker = Maps.newHashMap();
```

Comment:
This variable will be accessed by multiple threads in `sendShuffleDataAsync`.

Reply:
no, it's not shared since a new instance is created each time you call the method.

Reply:
I overlooked the CompletableFuture part inside sendShuffleDataAsync. let me rollback change.

Reply:
> no, it's not shared since a new instance is created each time you call the method.

You can see for more details.

https://github.com/apache/incubator-uniffle/blob/f4048fc9dc11b1da642e6e97e540ea98b57d490c/client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java#L172

Reply:
rolled back.

Reply:
I went through some logic and didn't find any update to "blockIdsTracker" (**correct me if I am wrong**) in main thread after "sendShuffleDataAsync" call which runs asynchronously in the threadpool, "dataTransferPool". According to  [BlockingQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html ) (used internally by the thread pool), "...actions in a thread prior to placing an object into a BlockingQueue [happen-before](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility) actions subsequent to the access or removal of that element from the BlockingQueue in another thread.". 

So, I think we don't need cocurrentHashmap for "blockIdsTracker".  And you use "AtomicInteger" as value part of "blockIdsTracker", it's enough to make the updated value visible to other threads in later code.

Reply:
You're right. But it's more safe to use `cocurrentHashmap`. If we modify this logic one day, we could forget to change this type to `ConcurrentHashmap`. If you still think it's  meaningful to modify this type, I think we could add some comments to explain why we don't use `ConcurrentHashmap` and remind us of this point.

Reply:
I just changed back to HashMap with comments to show reason. And from the code logic perspective, we will unlikely to insert/delete entries after dispatching it for sendShuffleDataAsync.